internal/agentcfg/elasticsearch.go

```diff
@@ -0,0 +1,237 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package agentcfg
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/elastic/elastic-agent-libs/monitoring"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+
+	"github.com/elastic/apm-server/internal/elasticsearch"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+)
+
+const ElasticsearchIndexName = ".apm-agent-configuration"
+
+type ElasticsearchFetcher struct {
+	client          elasticsearch.Client
+	cacheDuration   time.Duration
+	fallbackFetcher Fetcher
+
+	mu                  sync.RWMutex
+	last                time.Time
+	cache               []AgentConfig
+	expectedIndexExists bool
+
+	searchSize int
+
+	firstRunCompleted chan struct{}
+
+	logger *logp.Logger
+
+	metrics fetcherMetrics // TODO(carsonip): use pointer?
+}
+
+type fetcherMetrics struct {
+	fetchRequestCount, fetchNonFallback, fetchFallback, fetchFallbackMissing int64
+	refreshCacheSuccess, refreshCacheFailure                                 int64
+	entryCount                                                               int64
+}
+
+func NewElasticsearchFetcher(client elasticsearch.Client, cacheDuration time.Duration, fetcher Fetcher) *ElasticsearchFetcher {
+	logger := logp.NewLogger("agentcfg")
+	return &ElasticsearchFetcher{
+		client:            client,
+		cacheDuration:     cacheDuration,
+		fallbackFetcher:   fetcher,
+		searchSize:        1000, // TODO(carsonip): appropriate page size?
+		firstRunCompleted: make(chan struct{}),
+		logger:            logger,
+	}
+}
+
+// Fetch finds a matching agent config based on the received query.
+func (f *ElasticsearchFetcher) Fetch(ctx context.Context, query Query) (Result, error) {
+	atomic.AddInt64(&f.metrics.fetchRequestCount, 1)
+	select {
+	case <-ctx.Done():
+		return Result{}, ctx.Err()
+	case <-f.firstRunCompleted:
+	}
+
+	if !f.expectedIndexExists {
+		if f.fallbackFetcher == nil {
+			defer atomic.AddInt64(&f.metrics.fetchFallbackMissing, 1)
+			return Result{}, errors.New(ErrAgentRemoteConfigurationDisabled)
+		}
+		defer atomic.AddInt64(&f.metrics.fetchFallback, 1)
+		return f.fallbackFetcher.Fetch(ctx, query)
+	}
+
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	defer atomic.AddInt64(&f.metrics.fetchNonFallback, 1)
+	return matchAgentConfig(query, f.cache), nil
+}
+
+// Run refreshes the fetcher cache by querying Elasticsearch periodically.
+func (f *ElasticsearchFetcher) Run(ctx context.Context) error {
+	t := time.NewTicker(f.cacheDuration) // TODO(carsonip): cache duration
```

Comment:
single stale cache entries should generally be removed; if the whole cache update fails, I'd be fine with either - returning an error or returning last succesful entry.

Reply:
The entire cache is replaced on refresh cache. There will not be single stale cache entries. If the whole cache update fails, it will serve fetch requests from the stale cache, which is reasonable imo.

Reply:
Replacing the whole cache is not ideal allocation wise. The agent configurations are not expected to change that often, so only replacing changed settings would be more efficient. But we can tackle that in a follow up, to not block this PR from getting merged. Please create a github issue and add it to the `ms-8.7`.

Reply:
In order to not obtain the write lock during all the network calls, we are first loading everything into a buffer, and when all the queries and pagination requests are done, we obtain the write lock to the cache and reassign the cache variable to the buffer. i.e. during refresh cache, we will keep at max 2 copies of the agentconfigs in memory.

I do think there is room for a simple improvement here which is to preallocate the buffer by the size of the current cache. In most cases when the agent configs in ES are the same as what we have in our cache, we save allocations.

I can create a github issue to update the cache in-place.

Reply:
Tracked by #9930